### PR TITLE
MULE-13774: Add notifications to HTTP request

### PR DESCRIPTION
**Pull request by afelisatti**

### CODE REVIEW

File: src/main/java/org/mule/runtime/api/meta/model/ExtensionModel.java [hunk 1]
```diff
@@ -195,4 +197,14 @@
    * @return A {@link Set} of {@link ErrorModel} registered in the extension.
    */
   Set<ErrorModel> getErrorModels();
+
+  /**
+   * @return A {@link Set} of {@link NotificationModel} registered in the extension.
+   * @since 1.1
+   */
+  default Set<NotificationModel> getNotificationModels() {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/meta/model/ExtensionModel.java]: don't add a default. Add an entry in api-ignore instead.

File: src/main/java/org/mule/runtime/api/meta/model/notification/FiresNotifications.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Indicates that the current model can declare which {@link NotificationModel} it fires.
+ *
+ * @since 1.1
+ */
+public interface FiresNotifications {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/meta/model/notification/FiresNotifications.java]: I would rename to HasNotifications because 1) is more consistent 2) the model doesn't fire anything, the implementations do

File: src/main/java/org/mule/runtime/api/meta/model/notification/FiresNotifications.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.meta.model.notification;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Indicates that the current model can declare which {@link NotificationModel} it fires.
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/meta/model/notification/FiresNotifications.java]: copy paste the disclaimer I wrote in all the other models sayins that this interface is not to be implemented by the user

File: src/main/java/org/mule/runtime/api/meta/model/notification/FiresNotifications.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * @since 1.1
+ */
+public interface FiresNotifications {
+
+  /**
+   * @return a {@link Set} of {@link NotificationModel} with the notifications that the current component fires.
+   *
+   * @see NotificationModel
+   */
+  default Set<NotificationModel> getNotificationModels() {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/meta/model/notification/FiresNotifications.java]: no default. Plus, if you were to have a default, this should be emptySet() which is immutable

File: src/main/java/org/mule/runtime/api/meta/model/notification/ImmutableNotificationModel.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import static org.apache.commons.lang3.builder.EqualsBuilder.reflectionEquals;
+import static org.apache.commons.lang3.builder.HashCodeBuilder.reflectionHashCode;
+import org.mule.metadata.api.model.MetadataType;
+
+/**
+ * Default and immutable implementation of {@link NotificationModel}
+ *
+ * @since 1.1
+ */
+public class ImmutableNotificationModel implements NotificationModel {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/meta/model/notification/ImmutableNotificationModel.java]: why is this here instead of being in ext-api like the others?

Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/meta/model/notification/ImmutableNotificationModel.java]: final until proven otherwise

File: src/main/java/org/mule/runtime/api/meta/model/notification/ImmutableNotificationModel.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+  private final String namespace;
+  private final String identifier;
+  private final MetadataType metadataType;
+
+  public ImmutableNotificationModel(String namespace, String identifier, MetadataType metadataType) {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/meta/model/notification/ImmutableNotificationModel.java]: jdoc

File: src/main/java/org/mule/runtime/api/meta/model/notification/NotificationModel.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.meta.model.notification;
+
+import org.mule.metadata.api.model.MetadataType;
+import org.mule.runtime.api.meta.model.ComponentModel;
+
+/**
+ * A model which represents a notification that could be fired by the component represented by the {@link ComponentModel}.
+ *
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/meta/model/notification/NotificationModel.java]: disclaimer

File: src/main/java/org/mule/runtime/api/meta/model/notification/NotificationModel.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+   * @return the identifier for the notification
+   */
+  String getIdentifier();
+
+  /**
+   * Returns the {@link MetadataType} of the data associated to the notification.
+   *
+   * @return a not {@code null} {@link MetadataType}
+   */
+  MetadataType getDataType();
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/meta/model/notification/NotificationModel.java]: getDataType() -> getType()

File: src/main/java/org/mule/runtime/api/notification/Notification.java [hunk 9]
```diff
@@ -43,7 +43,21 @@ default boolean isSynchronous() {
   /**
    * Indicates the kind of action that causes a {@link Notification} to be fired.
    */
-  public interface Action {
+  interface Action {
+
+    /**
+     * @return unique namespace for the action
+     */
+    default String getNamespace() {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/notification/Notification.java]: I understand why you did this, but it makes no practical sense. Any Action implementation which doesn't override these methods will fail, not to mention that when you die and meet Liskov in heaven you won't be able to look her in the eye.

File: src/main/java/org/mule/runtime/api/meta/model/notification/ImmutableNotificationModel.java (hunk 5 above)
Comment (afelisatti) [reply src/main/java/org/mule/runtime/api/meta/model/notification/ImmutableNotificationModel.java]: because ErrorModel was my prototype

File: src/main/java/org/mule/runtime/api/meta/model/notification/NotificationModel.java (hunk 7 above)
Comment (afelisatti) [reply src/main/java/org/mule/runtime/api/meta/model/notification/NotificationModel.java]: Done.
